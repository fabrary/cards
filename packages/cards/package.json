{
  "name": "@flesh-and-blood/cards",
  "description": "TypeScript representations of Flesh and Blood cards",
  "version": "3.6.96",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "transform": "ts-node scripts/index.ts",
    "prebuild": "rm -rf dist",
    "build": "esbuild ./src/index.ts --outdir=dist --bundle --minify --platform=node --target=es6",
    "postbuild": "tsc --declaration && npm test",
    "test": "jest",
    "update:dependencies": "ncu -u latest-cards && npm i",
    "update:snapshots": "npm test -- -u",
    "new": "ts-node tests/new-cards.ts",
    "tts": "ts-node tts/index.ts",
    "sitemap": "ts-node sitemap/index.ts",
    "full": "npm run transform && npm run build && npm test && npm run new && npm run tts && npm run sitemap",
    "prepublish": "npm run full"
  },
  "dependencies": {
    "@flesh-and-blood/types": "^3.6.96"
  },
  "devDependencies": {
    "@types/jest": "^29.5.11",
    "@types/node": "^20.10.5",
    "@types/papaparse": "5.3.14",
    "chalk": "^4.1.2",
    "esbuild": "^0.19.10",
    "esbuild-jest": "^0.5.0",
    "fast-xml-parser": "4.3.2",
    "jest": "^29.7.0",
    "latest-cards": "npm:@flesh-and-blood/cards@^3.6.96",
    "papaparse": "^5.3.2",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3",
    "typo-js": "^1.2.3"
  },
  "files": [
    "/dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fabrary/fab-cards.git"
  },
  "bugs": {
    "url": "https://github.com/fabrary/fab-cards/issues"
  },
  "homepage": "https://github.com/fabrary/fab-cards#readme",
  "author": "Phillip Manwaring",
  "license": "MIT",
  "keywords": [
    "Flesh and Blood",
    "FAB",
    "FABTCG"
  ],
  "gitHead": "71162394e6b8ac0ef9a01cdaacc2661d4a34ae3e"
}
