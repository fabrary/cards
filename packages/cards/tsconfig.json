{
  "compilerOptions": {
    "emitDeclarationOnly": true,
    "esModuleInterop": true,
    "declaration": true,
    "isolatedModules": true,
    "lib": ["ES2021"],
    "module": "NodeNext",
    "moduleResolution": "nodenext",
    "outDir": "dist",
    "skipLibCheck": true,
    "target": "ES6"
  },
  "include": ["src"]
}
