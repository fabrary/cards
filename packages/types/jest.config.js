module.exports = {
  expand: true,
  // silent: true,
  testEnvironment: "node",
  testMatch: ["<rootDir>/tests/**.test.ts"],
  transform: {
    "^.+\\.tsx?$": "esbuild-jest",
  },
};
