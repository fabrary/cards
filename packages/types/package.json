{
  "name": "@flesh-and-blood/types",
  "description": "TypeScript representations of Flesh and Blood card types",
  "version": "0.0.62",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "prebuild": "rm -rf dist",
    "build": "esbuild ./src/index.ts --outdir=dist --bundle --minify --platform=node --target=es6",
    "postbuild": "tsc --declaration",
    "test": "jest",
    "full": "npm run test && npm run build",
    "prepublish": "npm run full"
  },
  "devDependencies": {
    "@types/jest": "^29.5.4",
    "@types/node": "^20.6.0",
    "@types/papaparse": "5.3.8",
    "chalk": "^4.1.2",
    "esbuild": "^0.19.2",
    "esbuild-jest": "^0.5.0",
    "fast-xml-parser": "^4.2.7",
    "jest": "^29.7.0",
    "papaparse": "5.3.2",
    "ts-node": "^10.9.1",
    "typescript": "^5.2.2"
  },
  "files": [
    "/dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fabrary/fab-cards.git"
  },
  "bugs": {
    "url": "https://github.com/fabrary/fab-cards/issues"
  },
  "homepage": "https://github.com/fabrary/fab-cards#readme",
  "author": "Phillip Manwaring",
  "license": "MIT",
  "keywords": [
    "Flesh and Blood",
    "FAB",
    "FABTCG"
  ],
  "gitHead": "999465737f8dcee7bebfcc5464ded359c87d24cf"
}
