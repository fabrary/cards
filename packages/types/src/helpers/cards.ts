import { Type, Keyword, Subtype } from "../interfaces";

export const getCardIdentifier = (
  card: {
    name: string;
    pitch?: string | number;
  },
  useNumber?: boolean
): string => {
  const { name: unformattedName, pitch } = card;
  const name = unformattedName
    .toLowerCase()
    .replace("//", "-")
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .replace(/ /g, "-")
    .replace("Ã­", "i")
    .replace(/[^a-z0-9 -]/g, "")
    .replace(/--/, "-");

  let suffix: string = "";
  switch (pitch) {
    case "1":
    case 1:
      suffix = useNumber ? "-1" : "-red";
      break;
    case "2":
    case 2:
      suffix = useNumber ? "-2" : "-yellow";
      break;
    case "3":
    case 3:
      suffix = useNumber ? "-3" : "-blue";
      break;
  }
  return `${name}${suffix}`;
};

export const getIsArenaCard = (types: Type[], keywords?: Keyword[]) => {
  const isDeckCard = getIsDeckCard(types, keywords);

  const isInventoryCard = [Type.DemiHero, Type.Equipment, Type.Weapon].some(
    (type) => types.includes(type)
  );

  return !isDeckCard && isInventoryCard;
};

export const getIsDeckCard = (types: Type[], keywords?: Keyword[]) => {
  const isDeckCardType = [
    Type.Action,
    Type.AttackReaction,
    Type.Block,
    Type.DefenseReaction,
    Type.Instant,
    Type.Mentor,
    Type.Resource,
  ].some((type) => types.includes(type));

  const isNotEphemeral = !keywords || !keywords.includes(Keyword.Ephemeral);

  return isDeckCardType && isNotEphemeral;
};

export const getCanAddToDeck = (types: Type[], keywords?: Keyword[]) => {
  const isArenaCard = getIsArenaCard(types, keywords);
  const isDeckCard = getIsDeckCard(types, keywords);

  return isArenaCard || isDeckCard;
};

export const getShouldRotateCardImage = (card: {
  keywords?: Keyword[];
  subtypes?: Subtype[];
}): boolean =>
  card.keywords?.includes(Keyword.Meld) ||
  card.subtypes?.includes(Subtype.Landmark) ||
  false;
