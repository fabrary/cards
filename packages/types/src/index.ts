export * from "./interfaces";
export * from "./sets";
