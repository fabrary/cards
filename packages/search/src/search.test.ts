import {
  DoubleSidedCard,
  Foiling,
  Hero,
  Meta,
  Rarity,
  Release,
  Treatment,
  Type,
} from "@flesh-and-blood/types";
import { cards } from "@flesh-and-blood/cards";
import Search from "./search";
import { setToSetIdentifierMappings } from "@flesh-and-blood/types";

const doubleSidedCards: DoubleSidedCard[] = cards.map((card) => {
  if (card.oppositeSideCardIdentifier) {
    const oppositeSideCard = cards.find(
      ({ cardIdentifier }) => cardIdentifier === card.oppositeSideCardIdentifier
    );
    if (oppositeSideCard) {
      (card as DoubleSidedCard).oppositeSideCard = oppositeSideCard;
    }
  }
  return card;
});

const exactSearches = [
  // Abbreviations & shorthands
  [3, "bbd"],
  [2, "cnc"],
  // [174, "poppers"],
  [1, "newminaris"],
  [1, "twominaris"],

  // Sets
  [237, "set:uprising", "set:upr", "s:uprising", "s:upr"],
  [53, 'set:"classic battles: rhinar vs dorinthea"', "s:dvr", "s:rvd"],
  [
    226,
    'set:"welcome to rathe"',
    "set:wtr",
    "set:rathe",
    'set:"welcome to"',
    's:"welcome to rathe"',
    "s:wtr",
    "s:rathe",
    's:"welcome to"',
  ],
  [236, "set:welcome", "s:welcome"], // Includes WTR and Ira Welcome deck
  [
    238,
    'set:"tales of aria"',
    "set:tales",
    "set:aria",
    "set:ele",
    's:"tales of aria"',
    "s:tales",
    "s:aria",
    "s:ele",
  ],

  // Popular searches
  [
    13,
    "set:uprising subtype:ally !type:placeholder",
    "s:uprising st:ally !t:placeholder",
  ],
  [
    45,
    'legal:fai talent:draconic keyword:"go again"',
    'l:fai tal:draconic k:"go again"',
  ],
  // [45, "type:equipment defense:>=2", "t:equipment d:>=2"],
  [27, 'class:wizard "arcane damage plus"', 'c:wizard "arcane damage plus"'],
  // [
  //   17,
  //   "legal:data class:mechanologist subtype:item cost:<=2",
  //   "l:data c:mechanologist st:item co:<=2",
  // ],
  // [66, "legal:rhinar power:>=6", "l:rhinar pwr:>=6"],
  [13, "legal:kassai copper", "l:kassai copper"],
  // [20, "legal:cc type:hero", "l:cc t:hero"],
  // [39, "legal:blitz type:hero", "l:blitz t:hero"],

  // Punctuation
  [1, "show time!", "show time"],
  [1, "seeker's leggings", "seekers leggings"],
  [1, "this round's on me", "this roundâ€™s on me", "this rounds on me"],

  // Text
  [19, "text:copper"],

  // Chain
  [10, "chain:dishonor"],
  [8, "chain:mugenshi"],

  // Referenced by
  [3, 'referencedBy:"Open the Center"'],
  [1, "referencedBy:Viserai"],

  // References
  [19, "references:copper"],
];

describe("Card search", () => {
  const cardSearch = new Search(doubleSidedCards);

  it.each(exactSearches)(
    "Gets %i cards for %s",
    (resultCount, ...searchTerms) => {
      for (const searchTerm of searchTerms) {
        const { searchResults } = cardSearch.search(
          randomizeCapitalization(searchTerm as string)
        );
        expect(searchResults.length).toBeGreaterThanOrEqual(
          resultCount as number
        );
      }
    }
  );

  interface Ratio {
    percent: number;
    lessThan?: boolean;
    searchTerms: string[];
  }
  const ratioSearches: (string | Ratio)[][] = [
    // Formats
    [
      "legal:blitz",
      {
        percent: 95,
        searchTerms: ["legal:blitz", "l:blitz"],
      },
    ],
    [
      "!legal:blitz",
      {
        percent: 5,
        lessThan: true,
        searchTerms: ["!legal:blitz", "-legal:blitz", "!l:blitz", "-l:blitz"],
      },
    ],
    [
      "legal:cc",
      {
        percent: 94,
        searchTerms: [
          'legal:"classic constructed"',
          "legal:cc",
          "legal:classic",
          "legal:constructed",
          'l:"classic constructed"',
          "l:cc",
          "l:classic",
          "l:constructed",
        ],
      },
    ],
    [
      "!legal:cc",
      {
        percent: 6,
        lessThan: true,
        searchTerms: ["!legal:cc", "-legal:cc", "!l:cc", "-l:cc"],
      },
    ],
    [
      "l:commoner",
      {
        percent: 50,
        searchTerms: ["legal:commoner", "l:commoner"],
      },
    ],
    [
      "!l:commoner",
      {
        percent: 50,
        lessThan: true,
        searchTerms: [
          "!legal:commoner",
          "-legal:commoner",
          "!l:commoner",
          "-l:commoner",
        ],
      },
    ],
    [
      "l:clash",
      {
        percent: 55,
        searchTerms: ["legal:clash", "l:clash"],
      },
    ],
    [
      "!l:clash",
      {
        percent: 45,
        lessThan: true,
        searchTerms: ["!legal:clash", "-legal:clash", "!l:clash", "-l:clash"],
      },
    ],

    // Classes
    [
      "c:generic",
      {
        percent: 10,
        searchTerms: ["class:generic", "c:generic"],
      },
    ],
  ];

  it.each(ratioSearches)("Gets card ratios for %s", (search, ratio) => {
    const { percent, lessThan, searchTerms } = ratio as Ratio;
    for (const searchTerm of searchTerms) {
      const { searchResults } = cardSearch.search(
        randomizeCapitalization(searchTerm as string)
      );
      const percentOfCards = Math.round(
        (100 * searchResults.length) / cards.length
      );
      if (lessThan) {
        expect(percentOfCards).toBeLessThan(percent);
      } else {
        expect(percentOfCards).toBeGreaterThan(percent);
      }
    }
  });

  const heroFilters = Object.values(Hero).map((hero) => [`l:"${hero}"`]);

  const hasQuantity: string[][] = [
    ...heroFilters,
    ["Kasaya"],
    ["Qi Unbound"],
    ["Ultron"],
    ["overloop"],
    ['t:"demi-hero"'],
    ["!pitch"],
    ["artist:delmiro"],
    ["treatment:ab"],
    ["treat:alt"],
    ["var:ea", "var:extended"],
    ['variation:"full art" foil:cold'],
    ["legal:blitz type:hero", "l:blitz t:hero"],
    ["legal:ira", "l:ira"],
    ["legal:emperor,clash cnc"],
    ["legal:shiyana", "l:shiyana"],
    ["legal:shiyana keyword:specialization", "l:shiyana k:specialization"],
    ["legal:shiyana,clash arknight ascendancy"],
    ["legal:prism,clash figment"],
    ["legal:boltyn"],
    ["legal:emperor"],
    ["legal:emperor cnc"],
    ["legal:iyslander"],
    ["legal:dromai"],
    ["legal:fai"],
    ["legal:nuu class:assassin"],
    ["legal:zen class:ninja"],
    ["legal:boltyn keyword:charge", "l:boltyn k:charge"],
    [
      'legal:iyslander pitch:3 type:action,instant !type:reaction !subtype:attack !keyword:"go again"',
      'l:iyslander p:3 t:action,instant !t:reaction !st:attack !k:"go again"',
    ],
    [
      'legal:benji power:<=2 keyword:"go again"',
      'l:benji pow:<=2 k:"go again"',
    ],
    ["legal:briar,clash rosetta thorn"],
    ,
  ];
  it.each(hasQuantity)("Gets cards for %s", (...searchTerms) => {
    for (const searchTerm of searchTerms) {
      const { keywords, searchResults, appliedFilters, attributes } =
        cardSearch.search(randomizeCapitalization(searchTerm as string));

      // console.log(
      //   JSON.stringify({ keywords, appliedFilters, attributes }, null, 2)
      // );

      expect(searchResults.length).toBeGreaterThan(0);
    }
  });

  const heroAndFirstClassFilters: string[] = cards
    .filter(({ types }) => types.includes(Type.Hero))
    .filter(({ hero }) => hero !== Hero.Taylor)
    .map(({ classes, hero }) => `l:"${hero}" c:"${classes[0]}"`);
  it.each(heroAndFirstClassFilters)("Gets cards for %s", (searchTerm) => {
    const { searchResults } = cardSearch.search(
      randomizeCapitalization(searchTerm as string)
    );
    expect(searchResults.length).toBeGreaterThan(0);
  });

  const artTreatmentFilters: string[] = Object.values(Treatment).map(
    (treatment) => `treatment:"${treatment}"`
  );
  it.each(artTreatmentFilters)("Gets cards for %s", (searchTerm) => {
    const { searchResults, appliedFilters } = cardSearch.search(
      randomizeCapitalization(searchTerm as string)
    );
    expect(searchResults.length).toBeGreaterThan(0);
  });

  const hasNoQuantity = [
    [
      "legal:ira,clash zephyr needle",
      "legal:shiyana,commoner cnc",
      "legal:shiyana,clash figment",
      "legal:shiyana,clash cnc",
      "legal:prism,commoner figment of erudition",
      "legal:prism,clash dread scythe",
      "legal:chane,clash rosetta thorn",
      "legal:shiyana,clash rosetta thorn",
    ],
  ];
  it.each(hasNoQuantity)("Gets zero cards for %s", (...searchTerms) => {
    for (const searchTerm of searchTerms) {
      const { searchResults } = cardSearch.search(
        randomizeCapitalization(searchTerm as string)
      );
      expect(searchResults.length).toEqual(0);
    }
  });

  const firstCards = [["lightning press", "lightning-press-red"]];
  it.each(firstCards)(
    "Gets first card first",
    (criteria, firstCardIdentifier) => {
      const { searchResults } = cardSearch.search(criteria);

      expect(searchResults[0].cardIdentifier).toEqual(firstCardIdentifier);
    }
  );

  it("Excludes rarities correctly", () => {
    const { searchResults } = cardSearch.search("!r:common, set:out");

    const cardsWithCommonRarity = searchResults
      .filter(({ rarities }) => rarities.includes(Rarity.Common))
      .map(({ name }) => name);
    expect(cardsWithCommonRarity).toEqual([]);
  });

  xit("Specific test", () => {
    const { searchResults, appliedFilters, keywords } = cardSearch.search(
      randomizeCapitalization("s:ros l:draft")
    );

    expect(searchResults.length).toBeGreaterThan(0);
    // expect(searchResults.length).toBe(0);
  });

  it("Combines heroes correctly", () => {
    const { searchResults, appliedFilters } = cardSearch.search(
      "l:zen,nuu s:mst r:t,c,r"
    );

    expect(searchResults.length).toEqual(148);
  });

  const punctuationCards = ["power:*", "{r}", "{p}"];

  it.each(punctuationCards)(
    "Handles punctuation correctly for %s",
    (criteria) => {
      const { searchResults } = cardSearch.search(criteria);

      expect(searchResults.length).toBeGreaterThan(0);
      expect(searchResults.length).toBeLessThan(Math.round(cards.length / 2));
    }
  );
});

const randomizeCapitalization = (str: string) =>
  str.replace(/./g, (char) =>
    Math.random() < 0.5 ? char.toUpperCase() : char
  );

describe("Returns artist when included", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Asur Misoa", () => {
    const artist = "Asur Misoa";
    const {
      attributes: { artists },
    } = cardSearch.search(`art:"${artist}"`);
    expect(artists).toBeTruthy();
    expect(artists[0]).toEqual(artist.toLowerCase());
  });

  it("Othon Nikolaidis", () => {
    const artist = "Othon";
    const {
      attributes: { artists },
    } = cardSearch.search(`r:marvel artist:${artist}`);
    expect(artists).toBeTruthy();
    expect(artists[0]).toEqual(artist.toLowerCase());
  });
});

describe("Sorts results by set when included", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Outsiders", () => {
    const { searchResults } = cardSearch.search("set:out");
    expect(searchResults.map(({ name }) => name)).toMatchSnapshot();
  });
});

describe("Additional heroes", () => {
  const cardSearch = new Search(doubleSidedCards, ["Another" as Hero]);

  it("Another hero", () => {
    const { appliedFilters } = cardSearch.search('l:"Another"');
    expect(appliedFilters.length).toBe(1);
    expect(appliedFilters[0].values).toEqual(["another"]);
    expect(appliedFilters[0].filterToPropertyMapping.property).toEqual(
      "legalHeroes"
    );
  });
});

describe("Returns set when included", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Outsiders", () => {
    const {
      attributes: { releases },
    } = cardSearch.search("set:out");
    expect(releases).toBeTruthy();
    expect(releases[0]).toEqual(Release.Outsiders);
  });

  it("Tales of Aria plus", () => {
    const {
      attributes: { releases },
    } = cardSearch.search('set:"Tales of Aria" talent:earth');
    expect(releases).toBeTruthy();
    expect(releases[0]).toEqual(Release.TalesOfAria);
  });
});

describe("Every set has results", () => {
  const cardSearch = new Search(doubleSidedCards);

  it.each(Object.values(Release))("%s has results", (set: string) => {
    const { searchResults } = cardSearch.search(`s:"${set}"`);

    expect(searchResults.length).toBeGreaterThan(0);
  });

  it.each(Object.values(Release))(
    "%s has results from abbreviated set",
    (set: string) => {
      const abbreviations = setToSetIdentifierMappings[set] as string[];

      for (const abbreviation of abbreviations) {
        const { searchResults } = cardSearch.search(`s:${abbreviation}`);
        expect(searchResults.length).toBeGreaterThan(0);
      }
    }
  );
});

describe("Armory decks are distinct", () => {
  const cardSearch = new Search(doubleSidedCards);

  it.each(
    Object.values(Release).filter((release) =>
      release.toUpperCase().includes("ARMORY DECK")
    )
  )("%s has distinct results", (set: string) => {
    const { searchResults, keywords, appliedFilters } = cardSearch.search(
      `s:"${set}"`
    );
    const cardsNotInSet = searchResults.filter(
      ({ sets }) => !sets.includes(set as Release)
    );
    expect(cardsNotInSet.length).toBe(0);
  });
});

describe("Returns foiling when included", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Rainbow foil", () => {
    const {
      attributes: { foilings },
    } = cardSearch.search("foil:rainbow");
    expect(foilings).toBeTruthy();
    expect(foilings[0]).toEqual(Foiling.R);
  });

  it("Cold and gold foils", () => {
    const {
      attributes: { foilings },
    } = cardSearch.search("foil:gold,c");
    expect(foilings).toBeTruthy();
    expect(foilings.length).toEqual(2);
  });
});

describe("Returns matching prints when release or foiling included", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Matching printing from foil", () => {
    const { searchResults } = cardSearch.search("foil:g");
    const searchResultsWithGoldFoilImage = searchResults.filter((card) =>
      card.printings.find(
        (printing) => printing.foiling === Foiling.G && !!printing.image
      )
    );
    for (const card of searchResultsWithGoldFoilImage) {
      for (const printing of card.matchingPrintings) {
        expect(printing).toBeTruthy();
        expect(printing.foiling).toEqual(Foiling.G);
      }
    }
  });

  it("Matching printing from set", () => {
    const { searchResults } = cardSearch.search("set:wtr");
    for (const card of searchResults) {
      for (const printing of card.matchingPrintings) {
        expect(printing).toBeTruthy();
        expect(printing.set).toEqual(Release.WelcomeToRathe);
      }
    }
  });

  it("Matching printing from treatment", () => {
    const { searchResults } = cardSearch.search("treat:aa");
    const searchResultsWithAltArtImage = searchResults;

    for (const card of searchResultsWithAltArtImage) {
      for (const printing of card.matchingPrintings) {
        expect(printing).toBeTruthy();
        expect(printing.treatment).toEqual(Treatment.AA);
      }
    }
  });

  it("Matching printing from rarity", () => {
    const { searchResults } = cardSearch.search("r:v");
    const searchResultsWithMarvelImages = searchResults;

    for (const card of searchResultsWithMarvelImages) {
      for (const printing of card.matchingPrintings) {
        expect(printing).toBeTruthy();
        expect(printing.rarity).toEqual(Rarity.Marvel);
      }
    }
  });

  it("Matching printing from card overlay limited filters", () => {
    const { searchResults } = cardSearch.search(
      "s:hnt !r:legendary,fabled l:draft,sealed"
    );
    for (const card of searchResults) {
      expect(card.matchingPrintings).toBeTruthy();
      if (!card.meta || !card.meta.includes(Meta.Expansion)) {
        expect(card.matchingPrintings.length).toBeGreaterThanOrEqual(1);
      }
      for (const printing of card.matchingPrintings) {
        expect(printing).toBeTruthy();
        expect(printing.set).toEqual(Release.TheHunted);
      }
    }
  });
});

describe("Shorthands property works", () => {
  const cardSearch = new Search(doubleSidedCards);

  it("Tap: leg tap", () => {
    const { searchResults } = cardSearch.search("tap");
    expect(searchResults).toBeTruthy();

    const legTap = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "leg-tap-red"
    );
    expect(legTap).toBeTruthy();
  });

  it("Tap: spitfire", () => {
    const { searchResults } = cardSearch.search("tap");
    expect(searchResults).toBeTruthy();

    const spitfire = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "spitfire"
    );
    expect(spitfire).toBeTruthy();
  });

  it("Leg Tap", () => {
    const { searchResults } = cardSearch.search("leg tap");
    expect(searchResults).toBeTruthy();

    const legTap = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "leg-tap-red"
    );
    expect(legTap).toBeTruthy();

    const spitfire = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "spitfire"
    );
    expect(spitfire).toBeFalsy();
  });

  it("Throw: throw caution to the wind", () => {
    const { searchResults, appliedFilters, keywords } =
      cardSearch.search("throw");
    expect(searchResults).toBeTruthy();

    console.log(JSON.stringify({ appliedFilters, keywords }, null, 2));
    expect(searchResults.length).toBeLessThan(doubleSidedCards.length);

    const throwCaution = searchResults.find(
      ({ cardIdentifier }) =>
        cardIdentifier === "throw-caution-to-the-wind-blue"
    );
    expect(throwCaution).toBeTruthy();
  });

  it("Throw: flick knives", () => {
    const { searchResults, appliedFilters, keywords } =
      cardSearch.search("throw");
    expect(searchResults).toBeTruthy();

    console.log(JSON.stringify({ appliedFilters, keywords }, null, 2));
    expect(searchResults.length).toBeLessThan(doubleSidedCards.length);

    const flickKnives = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "flick-knives"
    );
    expect(flickKnives).toBeTruthy();
  });

  it("Flick: flicker wisp", () => {
    const { searchResults } = cardSearch.search("flick");
    expect(searchResults).toBeTruthy();
    expect(searchResults.length).toBeLessThan(doubleSidedCards.length);

    const flickerWisp = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "flicker-wisp-yellow"
    );
    expect(flickerWisp).toBeTruthy();
  });

  it("Flick: flick knives", () => {
    const { searchResults } = cardSearch.search("flick");
    expect(searchResults).toBeTruthy();
    expect(searchResults.length).toBeLessThan(doubleSidedCards.length);

    const flickKnives = searchResults.find(
      ({ cardIdentifier }) => cardIdentifier === "flick-knives"
    );
    expect(flickKnives).toBeTruthy();
  });
});
