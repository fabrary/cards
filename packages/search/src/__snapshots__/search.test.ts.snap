// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorts results by set when included Outsiders 1`] = `
[
  "Plague Hive",
  "Uzuri, Switchblade",
  "Uzuri",
  "Arakni, Solitary Confinement",
  "Spider's Bite",
  "Nerve Scalpel",
  "Orbitoclast",
  "Scale Peeler",
  "Redback Shroud",
  "Infiltrate",
  "Shake Down",
  "Spreading Plague",
  "Back Stab",
  "Back Stab",
  "Back Stab",
  "Sneak Attack",
  "Sneak Attack",
  "Sneak Attack",
  "Spike with Bloodrot",
  "Spike with Frailty",
  "Spike with Inertia",
  "Infect",
  "Infect",
  "Infect",
  "Isolate",
  "Isolate",
  "Isolate",
  "Malign",
  "Malign",
  "Malign",
  "Prowl",
  "Prowl",
  "Prowl",
  "Sedate",
  "Sedate",
  "Sedate",
  "Wither",
  "Wither",
  "Wither",
  "Razor's Edge",
  "Razor's Edge",
  "Razor's Edge",
  "Katsu, the Wanderer",
  "Katsu",
  "Benji, the Piercing Wind",
  "Harmonized Kodachi",
  "Mask of Many Faces",
  "Cyclone Roundhouse",
  "Dishonor",
  "Head Leads the Tail",
  "Wander With Purpose",
  "Silverwind Shuriken",
  "Visit the Floating Dojo",
  "Bonds of Ancestry",
  "Bonds of Ancestry",
  "Bonds of Ancestry",
  "Recoil",
  "Recoil",
  "Recoil",
  "Spinning Wheel Kick",
  "Spinning Wheel Kick",
  "Spinning Wheel Kick",
  "Back Heel Kick",
  "Back Heel Kick",
  "Back Heel Kick",
  "Be Like Water",
  "Be Like Water",
  "Be Like Water",
  "Deadly Duo",
  "Deadly Duo",
  "Deadly Duo",
  "Descendent Gustwave",
  "Descendent Gustwave",
  "Descendent Gustwave",
  "Head Jab",
  "Head Jab",
  "Head Jab",
  "One-Two Punch",
  "One-Two Punch",
  "One-Two Punch",
  "Surging Strike",
  "Surging Strike",
  "Surging Strike",
  "Twin Twisters",
  "Twin Twisters",
  "Twin Twisters",
  "Azalea, Ace in the Hole",
  "Azalea",
  "Riptide, Lurker of the Deep",
  "Riptide",
  "Barbed Castaway",
  "Trench of Sunken Treasure",
  "Quiver of Abyssal Depths",
  "Quiver of Rustling Leaves",
  "Crow's Nest",
  "Driftwood Quiver",
  "Wayfinder's Crest",
  "Amplifying Arrow",
  "Barbed Undertow",
  "Buzzsaw Trap",
  "Collapsing Trap",
  "Spike Pit Trap",
  "Melting Point",
  "Boulder Trap",
  "Pendulum Trap",
  "Tarpit Trap",
  "Fletch a Red Tail",
  "Fletch a Yellow Tail",
  "Fletch a Blue Tail",
  "Lace with Bloodrot",
  "Lace with Frailty",
  "Lace with Inertia",
  "Falcon Wing",
  "Falcon Wing",
  "Falcon Wing",
  "Infecting Shot",
  "Infecting Shot",
  "Infecting Shot",
  "Murkmire Grapnel",
  "Murkmire Grapnel",
  "Murkmire Grapnel",
  "Sedation Shot",
  "Sedation Shot",
  "Sedation Shot",
  "Skybound Shot",
  "Skybound Shot",
  "Skybound Shot",
  "Spire Sniping",
  "Spire Sniping",
  "Spire Sniping",
  "Widowmaker",
  "Widowmaker",
  "Widowmaker",
  "Withering Shot",
  "Withering Shot",
  "Withering Shot",
  "Flick Knives",
  "Mask of Shifting Perspectives",
  "Blade Cuff",
  "Stab Wound",
  "Concealed Blade",
  "Knives Out",
  "Bleed Out",
  "Bleed Out",
  "Bleed Out",
  "Hurl",
  "Hurl",
  "Hurl",
  "Plunge",
  "Plunge",
  "Plunge",
  "Short and Sharp",
  "Short and Sharp",
  "Short and Sharp",
  "Mask of Malicious Manifestations",
  "Toxic Tips",
  "Codex of Bloodrot",
  "Codex of Frailty",
  "Codex of Inertia",
  "Death Touch",
  "Death Touch",
  "Death Touch",
  "Toxicity",
  "Toxicity",
  "Toxicity",
  "Virulent Touch",
  "Virulent Touch",
  "Virulent Touch",
  "Bloodrot Trap",
  "Frailty Trap",
  "Inertia Trap",
  "Vambrace of Determination",
  "Seeker's Hood",
  "Seeker's Gilet",
  "Seeker's Mitts",
  "Seeker's Leggings",
  "Silken Gi",
  "Threadbare Tunic",
  "Fisticuffs",
  "Fleet Foot Sandals",
  "Amnesia",
  "Down and Dirty",
  "Give and Take",
  "Gore Belching",
  "Burdens of the Past",
  "Premeditate",
  "Humble",
  "Humble",
  "Humble",
  "Infectious Host",
  "Infectious Host",
  "Infectious Host",
  "Looking for a Scrap",
  "Looking for a Scrap",
  "Looking for a Scrap",
  "Wreck Havoc",
  "Wreck Havoc",
  "Wreck Havoc",
  "Cut Down to Size",
  "Cut Down to Size",
  "Cut Down to Size",
  "Destructive Deliberation",
  "Destructive Deliberation",
  "Destructive Deliberation",
  "Feisty Locals",
  "Feisty Locals",
  "Feisty Locals",
  "Freewheeling Renegades",
  "Freewheeling Renegades",
  "Freewheeling Renegades",
  "Ravenous Rabble",
  "Ravenous Rabble",
  "Ravenous Rabble",
  "Seek Horizon",
  "Seek Horizon",
  "Seek Horizon",
  "Spring Load",
  "Spring Load",
  "Spring Load",
  "Come to Fight",
  "Come to Fight",
  "Come to Fight",
  "Scout the Periphery",
  "Scout the Periphery",
  "Scout the Periphery",
  "Brush Off",
  "Brush Off",
  "Brush Off",
  "Peace of Mind",
  "Peace of Mind",
  "Peace of Mind",
  "Bloodrot Pox",
  "Frailty",
  "Inertia",
  "Ponder",
  "Cracked Bauble",
]
`;
